File: epikjjh/baekjoon/2042.py
import math
import sys


def init(arr: list, tree: list, node: int, start: int, end: int)-> int:
    if start == end:
        tree[node] = arr[start]
        return tree[node]
    else:
        tree[node] = init(arr, tree, node*2, start, (start+end)//2) + init(arr, tree, node*2+1, (start+end)//2+1, end)
        return tree[node]
    
    
def sum(tree: list, node: int, start: int, end: int, left: int, right: int)-> int:
    if right < start or end < left:
        return 0
    elif left <= start and end <= right:
        return tree[node]
    else:
        return sum(tree, node*2, start, (start+end)//2, left, right) + sum(tree, node*2+1, (start+end)//2+1, end, left, right)
    
    
def update(tree: list, node: int, start: int, end: int, index: int, diff: int):
    if index < start or index > end:
        return
    tree[node] += diff
    if start != end:
        update(tree, node*2, start, (start+end)//2, index, diff)
        update(tree, node*2+1, (start+end)//2+1, end, index, diff)
        

input = sys.stdin.readline
num, m, k = map(int, input().rstrip().split())
arr = [int(input().rstrip()) for i in range(num)]
height = math.ceil(math.log2(num))
tree = [0]*2**(height+1)
init(arr, tree, 1, 0, num-1)
for i in range(m+k):
    a, b, c = map(int, input().rstrip().split())
    if a == 1:
        diff = c - (arr[b-1])
        arr[b-1] = c
        update(tree, 1, 0, num-1, b-1, diff)
    else:
        print(sum(tree, 1, 0, num-1, b-1, c-1))

File: epikjjh/baekjoon/2042.cpp
#include <bits/stdc++.h>

using namespace std;

const int SIZE = 2097152;

struct segTree{
	long long tree[SIZE];
	int size,leaf_start;
	segTree(int n): size(n){
		leaf_start = 1;
		while(leaf_start<size)	leaf_start *= 2;
		memset(tree,0,sizeof(tree));
	}
	void build_node(){
		for(int i=leaf_start-1;i>0;i--)	tree[i] = tree[i*2]+tree[i*2+1];
	}
	long long sum(int l, int r){
		return sum(1,l,r,0,leaf_start-1);
	}
	long long sum(int node, int l, int r, int n_l, int n_r){
		if(l<=n_l && n_r<=r)	return tree[node];
		if(r<n_l || n_r<l)	return 0;
		int mid = (n_l+n_r)/2;
		return sum(node*2,l,r,n_l,mid)+sum(node*2+1,l,r,mid+1,n_r);
	}
	void update(int index, long long value){
		int st=leaf_start+index;
		long long diff = value-tree[st];
		while(st>0){
			tree[st] += diff;
			st /= 2;
		}
	}
};

int main(){
	int n,m,k,a,b;
	long long c;
	scanf("%d %d %d", &n, &m, &k);
	segTree st(n);
	for(int i=0;i<n;i++)	scanf("%lld",&st.tree[st.leaf_start+i]);
	st.build_node();
	for(int i=0;i<m+k;i++){
		scanf("%d %d %lld", &a, &b, &c);
		if(a==1)	st.update(b-1,c);
		else		printf("%lld\n",st.sum(b-1,c-1));
	}
	return 0;
}